### ISSUE TITLE
Azure Active Directory based Authentication

### ISSUE THREAD

**Issue by davidhcar** [OHDSI/Atlas#2855]
Is it possible to use Azure based Active Directory Solution ? The closest configuration I see is this, 
```

	configLocal.userAuthenticationEnabled = true;
	configLocal.authProviders = [
		{
			"name": "Active Directory LDAP",
			"url": "user/login/ad",
			"ajax": true,
			"icon": "fa fa-cubes",
			"isUseCredentialsForm": false
		}];

```
However, the configurations for AAD need more than the URL parameter, see below configurations(https://learn.microsoft.com/en-us/azure/active-directory/develop/tutorial-v2-javascript-spa#configure-the-javascript-spa)

```
const msalConfig = {
    auth: {
      clientId: "Enter_the_Application_Id_Here",
      authority: "Enter_the_Cloud_Instance_Id_Here/Enter_the_Tenant_Info_Here",
      redirectUri: "Enter_the_Redirect_URI_Here",
    },
    cache: {
      cacheLocation: "sessionStorage", // This configures where your cache will be stored
      storeAuthStateInCookie: false, // Set this to "true" if you're having issues on Internet Explorer 11 or Edge
    }
  };

  // Add scopes for the ID token to be used at Microsoft identity platform endpoints.
  const loginRequest = {
    scopes: ["openid", "profile", "User.Read"]
  };

  // Add scopes for the access token to be used at Microsoft Graph API endpoints.
  const tokenRequest = {
    scopes: ["Mail.Read"]
  };
```

Could someone please provide guidance on configuration to use Atlas using Azure Active Directory. If there is a sample code, That would be great! Also, How to configure WebAPI to align with AAD based accesses.

**Comment by davidhcar** [comment]
Looks like I need to use either OpenID (OAUTH2 /authorize) or SAML ?

**Comment by konstjar** [comment]
You have the following options depending on AD type:

- On-Premises AD: LDAP connection, ADFS (OpenID, SAML)
- Azure AD: OpenID, SAML

Most of the configuration, you should do on WebAPI side. See the documentation:
https://github.com/OHDSI/WebAPI/wiki/Security-Configuration#webapi-security-configuration

**Comment by davidhcar** [comment]
Thank you @konstjar. The URL of authProviders should be the same as API `<security.oid.url>`  Is that correct ? and Could this url in Atlas UI be empty? as this is coming from api

```
configLocal.userAuthenticationEnabled = true;
	configLocal.authProviders = [
		{
			"name": "OpenID",
			"url": "user/login/ad",
			"ajax": true,
			"icon": "fa fa-cubes",
			"isUseCredentialsForm": false
		}];
```

**Comment by davidhcar** [comment]
When I click signinurl on the UI, it is going to `https://xxxxxxx/WebAPI/https://login.microsoftonline.com/xxxxx/v2.0/.well-known/openid-configuration?redirectUrl=/home`
It does not look like the URL is correct, what am I missing here. I configured everything under OpenId section under `https://github.com/OHDSI/WebAPI/wiki/Security-Configuration#webapi-security-configuration`

**Comment by davidhcar** [comment]
Realized the url on the client side does not change. I got through the authentication flow working, however the UI is still not recognizing the token and still throws 401 for WebAPI/user/me endpoint. So the sign in link remain same and no other features can be accessed throughout the app that requires successful authentication. 

Appreciate an insight into this. Thank you in advance!

**Comment by davidhcar** [comment]
```
ERROR http-nio-8080-exec-3 org.ohdsi.webapi.shiro.filters.ExceptionHandlerFilter - [] - Error during filtering
javax.servlet.ServletException: org.pac4j.core.exception.TechnicalException: State parameter is different from the one sent in authentication request. Session expired or possible threat of cross-site request forgery
```
This is the error seen in the log

**Comment by konstjar** [comment]
As I see from your ATLAS configuration, the url is incorrect. In case of OpenID, you should use "user/login/openid".

```
           configLocal.authProviders = [
		{
			"name": "OpenID",
			"url": "user/login/openid",
			"ajax": true,
			"icon": "fa fa-cubes",
			"isUseCredentialsForm": false
		}];

**Comment by davidhcar** [comment]
Thank you @konstjar .It is OpenId in the url I am using, that was a typo in the comment. The AAD authentication flow is working, however the session is not established that is the error I posted in the previous post.

**Comment by anthonysena** [comment]
@davidhcar I think this issue is resolved based on reading OHDSI/WebAPI#2268 so I'm closing this out. If this is not correct, please re-open and let me know. Thanks!
